src/todolistediting.js

```diff
@@ -0,0 +1,307 @@
+import ListEditing from './listediting';
+import TodoListCheckCommand from './todolistcheckcommand';
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
+import {
+	modelViewInsertion,
+	modelViewTextInsertion,
+	dataModelViewInsertion,
+	dataModelViewTextInsertion,
+	dataViewModelCheckmarkInsertion,
+	modelViewChangeChecked,
+	modelViewChangeType
+} from './todolistconverters';
+
+import { findInRange } from './utils';
+
+/**
+ * The engine of the todo list feature. It handles creating, editing and removing todo lists and its items.
+ *
+ * It registers all functionalities of {@link module:list/listediting~ListEditing list editing plugin} and extends
+ * it by `'todoList'` command.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class TodoListEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ ListEditing ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const { editing, data, model } = editor;
+		const viewDocument = editing.view.document;
+
+		// Extend schema.
+		model.schema.extend( 'listItem', {
+			allowAttributes: [ 'todoListChecked' ]
+		} );
+
+		// Disallow todoListChecked attribute on other nodes than listItem with todo listType.
+		model.schema.addAttributeCheck( ( context, attributeName ) => {
+			const item = context.last;
+
+			if ( attributeName == 'todoListChecked' && item.name == 'listItem' && item.getAttribute( 'listType' ) != 'todo' ) {
+				return false;
+			}
+		} );
+
+		// Register commands.
+		editor.commands.add( 'todoList', new ListCommand( editor, 'todo' ) );
+		editor.commands.add( 'todoListCheck', new TodoListCheckCommand( editor ) );
+
+		// Define converters.
+		editing.downcastDispatcher.on(
+			'insert:listItem',
+			modelViewInsertion( model, listItem => this._handleCheckmarkChange( listItem ) ),
+			{ priority: 'high' }
+		);
+		editing.downcastDispatcher.on( 'insert:$text', modelViewTextInsertion, { priority: 'high' } );
+		data.downcastDispatcher.on( 'insert:listItem', dataModelViewInsertion( model ), { priority: 'high' } );
+		data.downcastDispatcher.on( 'insert:$text', dataModelViewTextInsertion, { priority: 'high' } );
+
+		editing.downcastDispatcher.on(
+			'attribute:listType:listItem',
+			modelViewChangeType( listItem => this._handleCheckmarkChange( listItem ) )
+		);
+		editing.downcastDispatcher.on(
+			'attribute:todoListChecked:listItem',
+			modelViewChangeChecked( listItem => this._handleCheckmarkChange( listItem ) )
+		);
+
+		data.upcastDispatcher.on( 'element:input', dataViewModelCheckmarkInsertion, { priority: 'high' } );
+
+		// Collect all view nodes that have changed and use it to check if checkmark UI element is going to
+		// be re-rendered. If yes than view post-fixer should verify view structure.
+		const changedViewNodes = new Set();
+
+		for ( const viewRoot of viewDocument.roots ) {
+			this.listenTo( viewRoot, 'change:children', ( evt, node ) => changedViewNodes.add( node ) );
+		}
```

Comment:
It may happen that one add a root dynamically. It is a feature people request very ofter and I believe we should be ready for it. So you should listen on new roots too.

Reply:
I was thinking about it, but I felt I'm the only one who is not against dynamic roots :)

Reply:
I am more and more convinced to accept them. It still needs some fixes (now when you dynamically remove a root it crashes the editor) but it could be a very useful feature.

Reply:
Done.